Projects/LinkedListIterator/LinkedListIterator/LinkedListIterator/SLList.h

```diff
@@ -0,0 +1,374 @@
+		}
+		else if(Index < Size)
+		{
+			FNode* Previous = GetNode(Index-1);
+
+			FNode* NewNode = new FNode();
+			NewNode->Element = NewElement;
+			NewNode->Next = Previous->Next;
+
+			Previous->Next = NewNode;
+
+			Size++;
+		}	
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Remove(const int Index)
+	{
+		if (Index < 0 || Index >= Size) //Bounds Checking
+		{
+			return;
+		}
+		else if (Index == 0) //Head
+		{
+			FNode* NewHead = Head->Next;
+			delete Head;
+			Head = NewHead;
+
+			if (Size == 1)
+			{
+				Tail = nullptr;
+			}
+		}
+		else
+		{
+			FNode* Previous = GetNode(Index - 1);
+			FNode* Current = Previous->Next;
+			Previous->Next = Current->Next;
+			delete Current;
+
+			if (Index == Size - 1)
+			{
+				Tail = Previous;
+			}
+		}
+		Size--;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	void Clear()
+	{
+		for (int i = 0; i < Size; i++)
+		{
+			FNode* Current = Head;
+			Head = Head->Next;
+			delete Current;
+		}
+		Tail = nullptr;
+		Size = 0;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//								TEMPLATES & PREDICATES
+	//-------------------------------------------------------------------------------------------
+	template<typename Pred>
+	void ForEach(const Pred& Predicate)
+	{
+		FNode* Current = Head;
+		for (int i = 0; i < Size; i++)
+		{
+			Predicate(Current->Element);
+			Current = Current->Next;
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	template<typename Pred>
+	T& FindByPredicate(const Pred& Predicate)
+	{
+		FNode* Current = Head;
+		for (int i = 0; i < Size; i++)
+		{
+			if (Predicate(Current->Element))
+			{
+				return &Current->Element;
+			}
+			Current = Current->Next;
+		}
+		return nullptr;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	template<typename Pred>
+	TSLList<T> FilterByPredicate(const Pred& Predicate)
+	{
+		TSLList<T> FilteredList;
+		FNode* Current = Head;
+		for (int i = 0; i < Size; i++)
+		{
+			if (Predicate(Current->Element))
+			{
+				FilteredList.AddTail(Current->Element);
+			}
+			Current = Current->Next;
+		}
+		return FilteredList;
+	}
+
+	//-------------------------------------------------------------------------------------------
+	template<typename Pred>
+	void RemoveAllByPredicate(const Pred& Predicate)
+	{
+		FNode* Current = Head;
+		for (int i = 0; i < Size; i++)
+		{
+			T Element = Current->Element;
+			Current = Current->Next;
+
+			if (Predicate(Current->Element))
+			{
+				Remove(i);
+				i--;
+			}
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------
+	//										ITERATOR
+	//-------------------------------------------------------------------------------------------
+	class FIterator
+	{
```

Comment:
you should reorder your class declaring the Iterator class to the beginning so you can use the iterators inside of your functions, maybe if you have the time could you prove it